File: tests/cut/test_embedded_intersection_curves.cpp
/**
 * @file test_embedded_intersection_curves.cpp
 * @brief Tests for embedded_intersection_curves
 *
 * Tests for:
 * - Intersection curve correctness
 * - Topology preservation (manifold, closed meshes)
 * - Volume and surface area preservation
 * - Both return variants (with/without curves)
 * - All dynamic/static mesh combinations
 *
 * Copyright (c) 2025 Ziga Sajovic, XLAB
 */

#include <catch2/catch_test_macros.hpp>
#include <catch2/catch_template_test_macros.hpp>
#include <catch2/matchers/catch_matchers_floating_point.hpp>
#include <trueform/trueform.hpp>
#include "type_traits.hpp"
#include "mesh_generators.hpp"
#include <cmath>

// =============================================================================
// Test 1: Overlapping Spheres - Basic Embedding
// =============================================================================

TEMPLATE_TEST_CASE("embedded_intersection_curves_overlapping_spheres", "[embedded_intersection_curves]",
    (tf::test::type_pair_dyn2<std::int32_t, double, false, false>),
    (tf::test::type_pair_dyn2<std::int32_t, double, true, false>),
    (tf::test::type_pair_dyn2<std::int32_t, double, false, true>),
    (tf::test::type_pair_dyn2<std::int32_t, double, true, true>),
    (tf::test::type_pair_dyn2<std::int64_t, double, false, false>),
    (tf::test::type_pair_dyn2<std::int64_t, double, true, false>),
    (tf::test::type_pair_dyn2<std::int64_t, double, false, true>),
    (tf::test::type_pair_dyn2<std::int64_t, double, true, true>))
{
    using index_t = typename TestType::index_type;
    using real_t = typename TestType::real_type;
    constexpr bool dyn1 = TestType::is_dynamic1;
    constexpr bool dyn2 = TestType::is_dynamic2;

    real_t radius = real_t(1);
    real_t separation = real_t(1);

    auto sphere1 = tf::test::maybe_as_dynamic<dyn1>(
        tf::make_sphere_mesh<index_t>(radius, 50, 50));
    tf::ensure_positive_orientation(sphere1.polygons());

    auto sphere2 = tf::test::maybe_as_dynamic<dyn2>(
        tf::make_sphere_mesh<index_t>(radius, 50, 50));
    tf::ensure_positive_orientation(sphere2.polygons());

    auto sphere2_transform = tf::make_transformation_from_translation(
        tf::make_vector(separation, real_t(0), real_t(0)));
    auto sphere2_frame = tf::make_frame(sphere2_transform);

    real_t original_volume = tf::signed_volume(sphere1.polygons());
    real_t original_area = tf::area(sphere1.polygons());

    auto result = tf::embedded_intersection_curves(
        sphere1.polygons(),
        sphere2.polygons() | tf::tag(sphere2_frame));

    // Topology: manifold and closed
    REQUIRE(tf::is_manifold(result.polygons()));
    REQUIRE(tf::is_closed(result.polygons()));

    // Volume preserved
    real_t result_volume = tf::signed_volume(result.polygons());
    REQUIRE(std::abs(result_volume - original_volume) < original_volume * real_t(0.01));

    // Surface area preserved
    real_t result_area = tf::area(result.polygons());
    REQUIRE(std::abs(result_area - original_area) < original_area * real_t(0.01));

    // More faces than original (some were split)
    REQUIRE(result.polygons().size() >= sphere1.polygons().size());
}

// =============================================================================
// Test 2: Overlapping Spheres - With Curves Return
// =============================================================================

TEMPLATE_TEST_CASE("embedded_intersection_curves_with_curves", "[embedded_intersection_curves]",
    (tf::test::type_pair_dyn2<std::int32_t, double, false, false>),
    (tf::test::type_pair_dyn2<std::int32_t, double, true, false>),
    (tf::test::type_pair_dyn2<std::int32_t, double, false, true>),
    (tf::test::type_pair_dyn2<std::int32_t, double, true, true>),
    (tf::test::type_pair_dyn2<std::int64_t, double, false, false>),
    (tf::test::type_pair_dyn2<std::int64_t, double, true, false>),
    (tf::test::type_pair_dyn2<std::int64_t, double, false, true>),
    (tf::test::type_pair_dyn2<std::int64_t, double, true, true>))
{
    using index_t = typename TestType::index_type;
    using real_t = typename TestType::real_type;
    constexpr bool dyn1 = TestType::is_dynamic1;
    constexpr bool dyn2 = TestType::is_dynamic2;

    real_t radius = real_t(1);
    real_t separation = real_t(1);

    auto sphere1 = tf::test::maybe_as_dynamic<dyn1>(
        tf::make_sphere_mesh<index_t>(radius, 50, 50));
    tf::ensure_positive_orientation(sphere1.polygons());

    auto sphere2 = tf::test::maybe_as_dynamic<dyn2>(
        tf::make_sphere_mesh<index_t>(radius, 50, 50));
    tf::ensure_positive_orientation(sphere2.polygons());

    auto sphere2_transform = tf::make_transformation_from_translation(
        tf::make_vector(separation, real_t(0), real_t(0)));
    auto sphere2_frame = tf::make_frame(sphere2_transform);

    auto [result, curves] = tf::embedded_intersection_curves(
        sphere1.polygons(),
        sphere2.polygons() | tf::tag(sphere2_frame),
        tf::return_curves);

    // Topology: manifold and closed
    REQUIRE(tf::is_manifold(result.polygons()));
    REQUIRE(tf::is_closed(result.polygons()));

    // Curves should form a closed loop (intersection circle)
    REQUIRE(curves.paths().size() == 1);
    REQUIRE(curves.points().size() > 0);

    // The curve should be closed (first point == last point in path)
    auto path = curves.paths().front();
    REQUIRE(path.front() == path.back());
}

// =============================================================================
// Test 3: Non-Overlapping Meshes - No Intersection
// =============================================================================

TEMPLATE_TEST_CASE("embedded_intersection_curves_non_overlapping", "[embedded_intersection_curves]",
    (tf::test::type_pair_dyn2<std::int32_t, double, false, false>),
    (tf::test::type_pair_dyn2<std::int32_t, double, true, false>),
    (tf::test::type_pair_dyn2<std::int32_t, double, false, true>),
    (tf::test::type_pair_dyn2<std::int32_t, double, true, true>),
    (tf::test::type_pair_dyn2<std::int64_t, double, false, false>),
    (tf::test::type_pair_dyn2<std::int64_t, double, true, false>),
    (tf::test::type_pair_dyn2<std::int64_t, double, false, true>),
    (tf::test::type_pair_dyn2<std::int64_t, double, true, true>))
{
    using index_t = typename TestType::index_type;
    using real_t = typename TestType::real_type;
    constexpr bool dyn1 = TestType::is_dynamic1;
    constexpr bool dyn2 = TestType::is_dynamic2;

    auto box1 = tf::test::maybe_as_dynamic<dyn1>(
        tf::make_box_mesh<index_t>(real_t(1), real_t(1), real_t(1)));
    tf::ensure_positive_orientation(box1.polygons());

    auto box2 = tf::test::maybe_as_dynamic<dyn2>(
        tf::make_box_mesh<index_t>(real_t(1), real_t(1), real_t(1)));
    tf::ensure_positive_orientation(box2.polygons());

    auto box2_transform = tf::make_transformation_from_translation(
        tf::make_vector(real_t(5), real_t(0), real_t(0)));
    auto box2_frame = tf::make_frame(box2_transform);

    std::size_t original_faces = box1.polygons().size();
    std::size_t original_points = box1.points().size();
    real_t original_volume = tf::signed_volume(box1.polygons());

    auto result = tf::embedded_intersection_curves(
        box1.polygons(),
        box2.polygons() | tf::tag(box2_frame));

    // Same face and point count (no intersection)
    REQUIRE(result.polygons().size() == original_faces);
    REQUIRE(result.points().size() == original_points);

    // Volume unchanged
    real_t result_volume = tf::signed_volume(result.polygons());
    REQUIRE(std::abs(result_volume - original_volume) < tf::epsilon<real_t>);

    // Topology preserved
    REQUIRE(tf::is_manifold(result.polygons()));
    REQUIRE(tf::is_closed(result.polygons()));

    // With curves variant: no curves
    auto [result2, curves] = tf::embedded_intersection_curves(
        box1.polygons(),
        box2.polygons() | tf::tag(box2_frame),
        tf::return_curves);

    REQUIRE(curves.paths().size() == 0);
    REQUIRE(curves.points().size() == 0);
}

// =============================================================================
// Test 4: Overlapping Boxes - Intersection Curve Topology
// =============================================================================

TEMPLATE_TEST_CASE("embedded_intersection_curves_overlapping_boxes", "[embedded_intersection_curves]",
    (tf::test::type_pair_dyn2<std::int32_t, double, false, false>),
    (tf::test::type_pair_dyn2<std::int32_t, double, true, false>),
    (tf::test::type_pair_dyn2<std::int32_t, double, false, true>),
    (tf::test::type_pair_dyn2<std::int32_t, double, true, true>),
    (tf::test::type_pair_dyn2<std::int64_t, double, false, false>),
    (tf::test::type_pair_dyn2<std::int64_t, double, true, false>),
    (tf::test::type_pair_dyn2<std::int64_t, double, false, true>),
    (tf::test::type_pair_dyn2<std::int64_t, double, true, true>))
{
    using index_t = typename TestType::index_type;
    using real_t = typename TestType::real_type;
    constexpr bool dyn1 = TestType::is_dynamic1;
    constexpr bool dyn2 = TestType::is_dynamic2;

    auto box1 = tf::test::maybe_as_dynamic<dyn1>(
        tf::make_box_mesh<index_t>(real_t(1), real_t(1), real_t(1)));
    tf::ensure_positive_orientation(box1.polygons());

    auto box2 = tf::test::maybe_as_dynamic<dyn2>(
        tf::make_box_mesh<index_t>(real_t(1), real_t(1), real_t(1)));
    tf::ensure_positive_orientation(box2.polygons());

    auto box2_transform = tf::make_transformation_from_translation(
        tf::make_vector(real_t(0.5), real_t(0), real_t(0)));
    auto box2_frame = tf::make_frame(box2_transform);

    real_t original_volume = tf::signed_volume(box1.polygons());
    real_t original_area = tf::area(box1.polygons());

    auto [result, curves] = tf::embedded_intersection_curves(
        box1.polygons(),
        box2.polygons() | tf::tag(box2_frame),
        tf::return_curves);

    // Topology preserved
    REQUIRE(tf::is_manifold(result.polygons()));
    REQUIRE(tf::is_closed(result.polygons()));

    // Volume preserved
    real_t result_volume = tf::signed_volume(result.polygons());
    REQUIRE(std::abs(result_volume - original_volume) < original_volume * real_t(0.01));

    // Surface area preserved
    real_t result_area = tf::area(result.polygons());
    REQUIRE(std::abs(result_area - original_area) < original_area * real_t(0.01));

    // Box-box intersection forms closed loop(s)
    REQUIRE(curves.paths().size() >= 1);
}

// =============================================================================
// Test 5: Curve Matches Boolean Intersection Curve
// =============================================================================

TEMPLATE_TEST_CASE("embedded_intersection_curves_matches_boolean", "[embedded_intersection_curves]",
    (tf::test::type_pair_dyn2<std::int32_t, double, false, false>),
    (tf::test::type_pair_dyn2<std::int64_t, double, false, false>))
{
    using index_t = typename TestType::index_type;
    using real_t = typename TestType::real_type;

    real_t radius = real_t(1);
    real_t separation = real_t(1);

    auto sphere1 = tf::make_sphere_mesh<index_t>(radius, 30, 30);
    tf::ensure_positive_orientation(sphere1.polygons());

    auto sphere2 = tf::make_sphere_mesh<index_t>(radius, 30, 30);
    tf::ensure_positive_orientation(sphere2.polygons());

    auto sphere2_transform = tf::make_transformation_from_translation(
        tf::make_vector(separation, real_t(0), real_t(0)));
    auto sphere2_frame = tf::make_frame(sphere2_transform);

    // Get curves from embedded_intersection_curves
    auto [embedded_result, embedded_curves] = tf::embedded_intersection_curves(
        sphere1.polygons(),
        sphere2.polygons() | tf::tag(sphere2_frame),
        tf::return_curves);

    // Get intersection from boolean
    auto [boolean_result, labels] = tf::make_boolean(
        sphere1.polygons(),
        sphere2.polygons() | tf::tag(sphere2_frame),
        tf::boolean_op::intersection);

    // Embedded mesh has intersection edges; boolean intersection is the lens
    // The intersection curve should have same number of points
    // (they trace the same intersection circle)
    REQUIRE(embedded_curves.points().size() > 0);

    // Both results should be valid topology
    REQUIRE(tf::is_manifold(embedded_result.polygons()));
    REQUIRE(tf::is_closed(embedded_result.polygons()));
    REQUIRE(tf::is_manifold(boolean_result.polygons()));
    REQUIRE(tf::is_closed(boolean_result.polygons()));
}

// =============================================================================
// Test 6: Nested Spheres - Inner Fully Inside Outer
// =============================================================================

TEMPLATE_TEST_CASE("embedded_intersection_curves_nested", "[embedded_intersection_curves]",
    (tf::test::type_pair_dyn2<std::int32_t, double, false, false>),
    (tf::test::type_pair_dyn2<std::int32_t, double, true, true>),
    (tf::test::type_pair_dyn2<std::int64_t, double, false, false>),
    (tf::test::type_pair_dyn2<std::int64_t, double, true, true>))
{
    using index_t = typename TestType::index_type;
    using real_t = typename TestType::real_type;
    constexpr bool dyn1 = TestType::is_dynamic1;
    constexpr bool dyn2 = TestType::is_dynamic2;

    real_t outer_radius = real_t(2);
    real_t inner_radius = real_t(1);

    auto outer_sphere = tf::test::maybe_as_dynamic<dyn1>(
        tf::make_sphere_mesh<index_t>(outer_radius, 40, 40));
    tf::ensure_positive_orientation(outer_sphere.polygons());

    auto inner_sphere = tf::test::maybe_as_dynamic<dyn2>(
        tf::make_sphere_mesh<index_t>(inner_radius, 30, 30));
    tf::ensure_positive_orientation(inner_sphere.polygons());

    std::size_t original_faces = outer_sphere.polygons().size();
    std::size_t original_points = outer_sphere.points().size();
    real_t original_volume = tf::signed_volume(outer_sphere.polygons());

    auto [result, curves] = tf::embedded_intersection_curves(
        outer_sphere.polygons(),
        inner_sphere.polygons(),
        tf::return_curves);

    // No surface intersection when fully nested
    REQUIRE(result.polygons().size() == original_faces);
    REQUIRE(result.points().size() == original_points);

    // Volume unchanged
    real_t result_volume = tf::signed_volume(result.polygons());
    REQUIRE(std::abs(result_volume - original_volume) < original_volume * real_t(0.01));

    // Topology preserved
    REQUIRE(tf::is_manifold(result.polygons()));
    REQUIRE(tf::is_closed(result.polygons()));

    // No intersection curves
    REQUIRE(curves.paths().size() == 0);
}

File: python/tests/test_embedded_intersection_curves.py
"""
Tests for embedded_intersection_curves

Copyright (c) 2025 Žiga Sajovic, XLAB
Licensed for noncommercial use under the PolyForm Noncommercial License 1.0.0.
Commercial licensing available via info@polydera.com.
https://github.com/polydera/trueform
"""

import sys
import numpy as np
import pytest
import trueform as tf


REAL_DTYPES = [np.float32, np.float64]
INDEX_DTYPES = [np.int32, np.int64]
MESH_TYPES = ['triangle', 'dynamic']


def prepare_mesh(mesh):
    """Build required structures for operations."""
    mesh.build_tree()
    mesh.build_face_membership()
    mesh.build_manifold_edge_link()
    return mesh


def make_mesh(faces, points, mesh_type='triangle'):
    """Create mesh with specified type."""
    if isinstance(faces, tf.OffsetBlockedArray):
        return tf.Mesh(faces, points)

    if mesh_type == 'dynamic':
        dyn_faces = tf.as_offset_blocked(faces)
        return tf.Mesh(dyn_faces, points)
    return tf.Mesh(faces, points)


def get_num_faces(faces):
    """Get number of faces for both ndarray and OffsetBlockedArray."""
    if isinstance(faces, tf.OffsetBlockedArray):
        return len(faces)
    return faces.shape[0]


# ==============================================================================
# Test 1: Overlapping Spheres - Basic Embedding
# ==============================================================================

@pytest.mark.parametrize("index_dtype", INDEX_DTYPES)
@pytest.mark.parametrize("dtype", REAL_DTYPES)
@pytest.mark.parametrize("mesh_type", MESH_TYPES)
def test_overlapping_spheres(index_dtype, dtype, mesh_type):
    """Embed intersection curve between overlapping spheres."""
    radius = dtype(1.0)
    separation = dtype(1.0)

    faces1, points1 = tf.make_sphere_mesh(
        radius, stacks=50, segments=50, dtype=dtype, index_dtype=index_dtype)
    sphere1 = prepare_mesh(make_mesh(faces1, points1, mesh_type))

    faces2, points2 = tf.make_sphere_mesh(
        radius, stacks=50, segments=50, dtype=dtype, index_dtype=index_dtype)
    points2_translated = points2.copy()
    points2_translated[:, 0] += float(separation)
    sphere2 = prepare_mesh(make_mesh(faces2, points2_translated, mesh_type))

    original_volume = tf.volume((faces1, points1))
    original_area = tf.area((faces1, points1))

    result_faces, result_points = tf.embedded_intersection_curves(sphere1, sphere2)

    # Topology preserved
    result_mesh = tf.Mesh(result_faces, result_points)
    assert tf.is_manifold(result_mesh)
    assert tf.is_closed(result_mesh)

    # Volume preserved
    result_volume = tf.volume((result_faces, result_points))
    np.testing.assert_allclose(result_volume, original_volume, rtol=0.01)

    # Area preserved
    result_area = tf.area((result_faces, result_points))
    np.testing.assert_allclose(result_area, original_area, rtol=0.01)

    # More faces than original (some were split)
    assert get_num_faces(result_faces) >= get_num_faces(faces1)


# ==============================================================================
# Test 2: Overlapping Spheres - With Curves Return
# ==============================================================================

@pytest.mark.parametrize("index_dtype", INDEX_DTYPES)
@pytest.mark.parametrize("dtype", REAL_DTYPES)
@pytest.mark.parametrize("mesh_type", MESH_TYPES)
def test_overlapping_spheres_with_curves(index_dtype, dtype, mesh_type):
    """Embed intersection curve and return the curve geometry."""
    radius = dtype(1.0)
    separation = dtype(1.0)

    faces1, points1 = tf.make_sphere_mesh(
        radius, stacks=50, segments=50, dtype=dtype, index_dtype=index_dtype)
    sphere1 = prepare_mesh(make_mesh(faces1, points1, mesh_type))

    faces2, points2 = tf.make_sphere_mesh(
        radius, stacks=50, segments=50, dtype=dtype, index_dtype=index_dtype)
    points2_translated = points2.copy()
    points2_translated[:, 0] += float(separation)
    sphere2 = prepare_mesh(make_mesh(faces2, points2_translated, mesh_type))

    (result_faces, result_points), (paths, curve_points) = tf.embedded_intersection_curves(
        sphere1, sphere2, return_curves=True)

    # Topology preserved
    result_mesh = tf.Mesh(result_faces, result_points)
    assert tf.is_manifold(result_mesh)
    assert tf.is_closed(result_mesh)

    # Curves should form a closed loop (intersection circle)
    assert len(paths) == 1
    assert len(curve_points) > 0

    # The curve should be closed
    path = paths[0]
    assert path[0] == path[-1]


# ==============================================================================
# Test 3: Non-Overlapping Meshes - No Intersection
# ==============================================================================

@pytest.mark.parametrize("index_dtype", INDEX_DTYPES)
@pytest.mark.parametrize("dtype", REAL_DTYPES)
@pytest.mark.parametrize("mesh_type", MESH_TYPES)
def test_non_overlapping(index_dtype, dtype, mesh_type):
    """Non-overlapping meshes return mesh unchanged."""
    faces1, points1 = tf.make_box_mesh(1.0, 1.0, 1.0, dtype=dtype, index_dtype=index_dtype)
    box1 = prepare_mesh(make_mesh(faces1, points1, mesh_type))

    faces2, points2 = tf.make_box_mesh(1.0, 1.0, 1.0, dtype=dtype, index_dtype=index_dtype)
    points2_translated = points2.copy()
    points2_translated[:, 0] += 5.0
    box2 = prepare_mesh(make_mesh(faces2, points2_translated, mesh_type))

    original_num_faces = get_num_faces(faces1)
    original_num_points = points1.shape[0]
    original_volume = tf.volume((faces1, points1))

    result_faces, result_points = tf.embedded_intersection_curves(box1, box2)

    # Same face and point count
    assert get_num_faces(result_faces) == original_num_faces
    assert result_points.shape[0] == original_num_points

    # Volume unchanged
    result_volume = tf.volume((result_faces, result_points))
    np.testing.assert_allclose(result_volume, original_volume, rtol=0.001)

    # Topology preserved
    result_mesh = tf.Mesh(result_faces, result_points)
    assert tf.is_manifold(result_mesh)
    assert tf.is_closed(result_mesh)


@pytest.mark.parametrize("index_dtype", INDEX_DTYPES)
@pytest.mark.parametrize("dtype", REAL_DTYPES)
def test_non_overlapping_with_curves(index_dtype, dtype):
    """Non-overlapping meshes return no curves."""
    faces1, points1 = tf.make_box_mesh(1.0, 1.0, 1.0, dtype=dtype, index_dtype=index_dtype)
    box1 = prepare_mesh(make_mesh(faces1, points1))

    faces2, points2 = tf.make_box_mesh(1.0, 1.0, 1.0, dtype=dtype, index_dtype=index_dtype)
    points2_translated = points2.copy()
    points2_translated[:, 0] += 5.0
    box2 = prepare_mesh(make_mesh(faces2, points2_translated))

    (result_faces, result_points), (paths, curve_points) = tf.embedded_intersection_curves(
        box1, box2, return_curves=True)

    assert len(paths) == 0
    assert len(curve_points) == 0


# ==============================================================================
# Test 4: Overlapping Boxes
# ==============================================================================

@pytest.mark.parametrize("index_dtype", INDEX_DTYPES)
@pytest.mark.parametrize("dtype", REAL_DTYPES)
@pytest.mark.parametrize("mesh_type", MESH_TYPES)
def test_overlapping_boxes(index_dtype, dtype, mesh_type):
    """Embed intersection curves between overlapping boxes."""
    faces1, points1 = tf.make_box_mesh(1.0, 1.0, 1.0, dtype=dtype, index_dtype=index_dtype)
    box1 = prepare_mesh(make_mesh(faces1, points1, mesh_type))

    faces2, points2 = tf.make_box_mesh(1.0, 1.0, 1.0, dtype=dtype, index_dtype=index_dtype)
    points2_translated = points2.copy()
    points2_translated[:, 0] += 0.5
    box2 = prepare_mesh(make_mesh(faces2, points2_translated, mesh_type))

    original_volume = tf.volume((faces1, points1))
    original_area = tf.area((faces1, points1))

    (result_faces, result_points), (paths, curve_points) = tf.embedded_intersection_curves(
        box1, box2, return_curves=True)

    # Topology preserved
    result_mesh = tf.Mesh(result_faces, result_points)
    assert tf.is_manifold(result_mesh)
    assert tf.is_closed(result_mesh)

    # Volume preserved
    result_volume = tf.volume((result_faces, result_points))
    np.testing.assert_allclose(result_volume, original_volume, rtol=0.01)

    # Area preserved
    result_area = tf.area((result_faces, result_points))
    np.testing.assert_allclose(result_area, original_area, rtol=0.01)

    # Box-box intersection forms closed loop(s)
    assert len(paths) >= 1


# ==============================================================================
# Test 5: Nested Spheres - No Surface Intersection
# ==============================================================================

@pytest.mark.parametrize("index_dtype", INDEX_DTYPES)
@pytest.mark.parametrize("dtype", REAL_DTYPES)
@pytest.mark.parametrize("mesh_type", MESH_TYPES)
def test_nested_spheres(index_dtype, dtype, mesh_type):
    """Nested spheres have no surface intersection."""
    outer_radius = dtype(2.0)
    inner_radius = dtype(1.0)

    faces1, points1 = tf.make_sphere_mesh(
        outer_radius, stacks=40, segments=40, dtype=dtype, index_dtype=index_dtype)
    outer = prepare_mesh(make_mesh(faces1, points1, mesh_type))

    faces2, points2 = tf.make_sphere_mesh(
        inner_radius, stacks=30, segments=30, dtype=dtype, index_dtype=index_dtype)
    inner = prepare_mesh(make_mesh(faces2, points2, mesh_type))

    original_num_faces = get_num_faces(faces1)
    original_num_points = points1.shape[0]
    original_volume = tf.volume((faces1, points1))

    (result_faces, result_points), (paths, curve_points) = tf.embedded_intersection_curves(
        outer, inner, return_curves=True)

    # Same face and point count
    assert get_num_faces(result_faces) == original_num_faces
    assert result_points.shape[0] == original_num_points

    # Volume unchanged
    result_volume = tf.volume((result_faces, result_points))
    np.testing.assert_allclose(result_volume, original_volume, rtol=0.01)

    # Topology preserved
    result_mesh = tf.Mesh(result_faces, result_points)
    assert tf.is_manifold(result_mesh)
    assert tf.is_closed(result_mesh)

    # No intersection curves
    assert len(paths) == 0


# ==============================================================================
# Test 6: Mixed Index Types (int32 x int64 and int64 x int32)
# ==============================================================================

@pytest.mark.parametrize("dtype", REAL_DTYPES)
def test_mixed_index_types_int32_int64(dtype):
    """int32 mesh with int64 mesh works."""
    faces1, points1 = tf.make_sphere_mesh(1.0, stacks=30, segments=30, dtype=dtype, index_dtype=np.int32)
    sphere1 = prepare_mesh(tf.Mesh(faces1, points1))

    faces2, points2 = tf.make_sphere_mesh(1.0, stacks=30, segments=30, dtype=dtype, index_dtype=np.int64)
    points2[:, 0] += 1.0
    sphere2 = prepare_mesh(tf.Mesh(faces2, points2))

    result_faces, result_points = tf.embedded_intersection_curves(sphere1, sphere2)

    result_mesh = tf.Mesh(result_faces, result_points)
    assert tf.is_manifold(result_mesh)
    assert tf.is_closed(result_mesh)


@pytest.mark.parametrize("dtype", REAL_DTYPES)
def test_mixed_index_types_int64_int32(dtype):
    """int64 mesh with int32 mesh works (asymmetric case)."""
    faces1, points1 = tf.make_sphere_mesh(1.0, stacks=30, segments=30, dtype=dtype, index_dtype=np.int64)
    sphere1 = prepare_mesh(tf.Mesh(faces1, points1))

    faces2, points2 = tf.make_sphere_mesh(1.0, stacks=30, segments=30, dtype=dtype, index_dtype=np.int32)
    points2[:, 0] += 1.0
    sphere2 = prepare_mesh(tf.Mesh(faces2, points2))

    result_faces, result_points = tf.embedded_intersection_curves(sphere1, sphere2)

    result_mesh = tf.Mesh(result_faces, result_points)
    assert tf.is_manifold(result_mesh)
    assert tf.is_closed(result_mesh)


# ==============================================================================
# Test 7: Mixed Mesh Types
# ==============================================================================

@pytest.mark.parametrize("index_dtype", INDEX_DTYPES)
@pytest.mark.parametrize("dtype", REAL_DTYPES)
def test_mixed_mesh_types_triangle_dynamic(index_dtype, dtype):
    """Triangle mesh0 with dynamic mesh1 returns triangle result."""
    faces1, points1 = tf.make_box_mesh(1.0, 1.0, 1.0, dtype=dtype, index_dtype=index_dtype)
    box1 = prepare_mesh(make_mesh(faces1, points1, 'triangle'))

    faces2, points2 = tf.make_box_mesh(1.0, 1.0, 1.0, dtype=dtype, index_dtype=index_dtype)
    points2_translated = points2.copy()
    points2_translated[:, 0] += 0.5
    box2 = prepare_mesh(make_mesh(faces2, points2_translated, 'dynamic'))

    result_faces, result_points = tf.embedded_intersection_curves(box1, box2)

    # Result type depends only on mesh0 (triangle), not mesh1
    assert isinstance(result_faces, np.ndarray)

    result_mesh = tf.Mesh(result_faces, result_points)
    assert tf.is_manifold(result_mesh)
    assert tf.is_closed(result_mesh)


@pytest.mark.parametrize("index_dtype", INDEX_DTYPES)
@pytest.mark.parametrize("dtype", REAL_DTYPES)
def test_mixed_mesh_types_dynamic_triangle(index_dtype, dtype):
    """Dynamic mesh0 with triangle mesh1 returns dynamic result."""
    faces1, points1 = tf.make_box_mesh(1.0, 1.0, 1.0, dtype=dtype, index_dtype=index_dtype)
    box1 = prepare_mesh(make_mesh(faces1, points1, 'dynamic'))

    faces2, points2 = tf.make_box_mesh(1.0, 1.0, 1.0, dtype=dtype, index_dtype=index_dtype)
    points2_translated = points2.copy()
    points2_translated[:, 0] += 0.5
    box2 = prepare_mesh(make_mesh(faces2, points2_translated, 'triangle'))

    result_faces, result_points = tf.embedded_intersection_curves(box1, box2)

    # Result type depends only on mesh0 (dynamic), not mesh1
    assert isinstance(result_faces, tf.OffsetBlockedArray)

    result_mesh = tf.Mesh(result_faces, result_points)
    assert tf.is_manifold(result_mesh)
    assert tf.is_closed(result_mesh)


# ==============================================================================
# Test 8: Error Handling
# ==============================================================================

def test_rejects_non_mesh_input():
    """Rejects non-Mesh inputs."""
    faces = np.array([[0, 1, 2]], dtype=np.int32)
    points = np.array([[0, 0, 0], [1, 0, 0], [0, 1, 0]], dtype=np.float32)
    mesh = prepare_mesh(tf.Mesh(faces, points))

    with pytest.raises(TypeError, match="must be a Mesh"):
        tf.embedded_intersection_curves((faces, points), mesh)


def test_rejects_2d_meshes():
    """Rejects 2D meshes."""
    faces = np.array([[0, 1, 2]], dtype=np.int32)
    points_2d = np.array([[0, 0], [1, 0], [0, 1]], dtype=np.float32)
    mesh_2d = tf.Mesh(faces, points_2d)

    points_3d = np.array([[0, 0, 0], [1, 0, 0], [0, 1, 0]], dtype=np.float32)
    mesh_3d = prepare_mesh(tf.Mesh(faces, points_3d))

    with pytest.raises(ValueError, match="3D"):
        tf.embedded_intersection_curves(mesh_2d, mesh_3d)


def test_rejects_mismatched_dtypes():
    """Rejects meshes with different real dtypes."""
    faces = np.array([[0, 1, 2]], dtype=np.int32)

    mesh_f32 = prepare_mesh(tf.Mesh(faces, np.array([[0, 0, 0], [1, 0, 0], [0, 1, 0]], dtype=np.float32)))
    mesh_f64 = prepare_mesh(tf.Mesh(faces, np.array([[0, 0, 0], [1, 0, 0], [0, 1, 0]], dtype=np.float64)))

    with pytest.raises(ValueError, match="dtype"):
        tf.embedded_intersection_curves(mesh_f32, mesh_f64)


if __name__ == "__main__":
    sys.exit(pytest.main([__file__, "-v"]))
